src/main/java/soot/asm/AsmMethodSource.java

```diff
@@ -2239,72 +2240,107 @@ public Body getBody(SootMethod m, String phaseName) {
-              }
-              Local newLocal = newLocals.get(localNum, name);
-              if (newLocal == null) {
-                newLocal = jimp.newLocal(name, UnknownType.v());
-                jbLocals.add(newLocal);
-                Local old = newLocals.put(localNum, name, newLocal);
-                assert (old == null);
+      {
+        // Group LocalVariableNode by index to find any that are associated with
+        // different names at different points in the method. For each such
+        // occurrence, determine which name was chosen via "locals.get(i)" and,
+        // in the range of Units specified for all other names, replace that
+        // chosen Local with a new Local.
+        Multimap<Integer, LocalVariableNode> groups = LinkedListMultimap.create(this.localVars.size());
+        for (LocalVariableNode lvn : this.localVars) {
+          if (lvn.start != lvn.end) { // these are ignored by getLocal(int)
+            groups.put(lvn.index, lvn);
+          }
+        }
+        // NOTE: When creating new variables, group by both name and index because
+        // the LocalVariableTable allows multiple local variable indices to
+        // have the same name simultaneously but they must be distinguished here.
+        Table<Integer, String, Local> newLocals = null;
+        for (Map.Entry<Integer, Collection<LocalVariableNode>> e : groups.asMap().entrySet()) {
+          Collection<LocalVariableNode> vals = e.getValue();
+          if (vals.size() > 1) {
+            final Integer localNum = e.getKey();
+            final Local chosen = this.locals.get(localNum);
+            final String chosenName = chosen.getName();
+            final Type chosenType = chosen.getType();
+            // Detect inconsistencies in the LocalVariableTable.
+            // 1. If there exists any use of local variable 'chosen' outside of a
+            // range defined by one of the LocalVariableNode in 'vals', then it is
+            // not safe to make any replacements of 'chosen' because it is not
+            // clear which actual variable should be used at a location outside of
+            // the defined ranges (unless a use-def analysis is applied but that
+            // is left for future implementation).
+            // 2. If any of the LocalVariableNode in 'vals' cover any of the same
+            // units, then they are ambiguous and cannot be used.
+            //
+            // To implement these checks, first collect all ValueBoxes in the body
+            // that reference the chosen Local. Then, as each LocalVariableNode is
+            // processed, map each ValueBox to the new Local that it should hold.
+            // If any ValueBox is found more than once or not found at all, then
+            // one of the inconsistency cases mentioned above exists and thus no
+            // changes should be made.
+            IdentityHashMap<ValueBox, Local> boxToNewLoc = new IdentityHashMap<>();
```

Comment:
@tim-hoffman  Can you split he following piece of code into multiple methods? The getBody  method gets really long. For example, the code you check for consistency can be one method.

Reply:
I split it into a couple of methods. I did consider making the consistency check inside the loop into its own method but it would end up needing a lot of parameters for all of the local variables at that point so I thought it would be best to leave it.